File: app/src/main/java/org/autojs/autojs/external/ScriptIntents.java
package org.autojs.autojs.external;

import android.content.Context;
import android.content.Intent;

import org.autojs.autojs.AutoJs;
import org.autojs.autojs.execution.ExecutionConfig;
import org.autojs.autojs.model.script.PathChecker;
import org.autojs.autojs.script.JavaScriptFileSource;
import org.autojs.autojs.script.ScriptSource;
import org.autojs.autojs.script.SequenceScriptSource;
import org.autojs.autojs.script.StringScriptSource;
import org.autojs.autojs.util.WorkingDirectoryUtils;
import org.json.JSONObject;

import java.io.File;

/**
 * Created by Stardust on 2017/4/1.
 */
public class ScriptIntents {

    public static final String EXTRA_KEY_PATH = "path";
    public static final String EXTRA_KEY_PRE_EXECUTE_SCRIPT = "script";
    public static final String EXTRA_KEY_LOOP_TIMES = "loop";
    public static final String EXTRA_KEY_LOOP_INTERVAL = "interval";
    public static final String EXTRA_KEY_DELAY = "delay";

    public static boolean isTaskerJsonObjectValid(JSONObject json) {
        return json.has(ScriptIntents.EXTRA_KEY_PATH) || json.has(EXTRA_KEY_PRE_EXECUTE_SCRIPT);
    }

    public static boolean handleIntent(Context context, Intent intent) {
        String path = getPath(intent);
        String script = intent.getStringExtra(ScriptIntents.EXTRA_KEY_PRE_EXECUTE_SCRIPT);
        int loopTimes = intent.getIntExtra(EXTRA_KEY_LOOP_TIMES, 1);
        long delay = intent.getLongExtra(EXTRA_KEY_DELAY, 0);
        long interval = intent.getLongExtra(EXTRA_KEY_LOOP_INTERVAL, 0);
        ScriptSource source = null;
        ExecutionConfig config = new ExecutionConfig();
        config.setDelay(delay);
        config.setLoopTimes(loopTimes);
        config.setInterval(interval);
        config.setArgument("intent", intent);
        if (path == null && script != null) {
            source = new StringScriptSource(script);
        } else if (path != null && new PathChecker(context).checkAndToastError(path)) {
            JavaScriptFileSource fileScriptSource = new JavaScriptFileSource(path);
            if (script != null) {
                source = new SequenceScriptSource(fileScriptSource.getName(), new StringScriptSource(script), fileScriptSource);
            } else {
                source = fileScriptSource;
            }
            String workingDirectory = new File(path).getParent();
            if (workingDirectory != null) {
                config.setWorkingDirectory(workingDirectory);
            }
        } else {
            config.setWorkingDirectory(WorkingDirectoryUtils.getPath());
        }
        if (source == null) {
            return false;
        }
        AutoJs.getInstance().getScriptEngineService().execute(source, config);
        return true;
    }

    private static String getPath(Intent intent) {
        if (intent.getData() != null && intent.getData().getPath() != null)
            return intent.getData().getPath();
        return intent.getStringExtra(ScriptIntents.EXTRA_KEY_PATH);
    }
}

File: app/src/main/java/org/autojs/autojs/external/ScriptIntents.kt
package org.autojs.autojs.external

import android.content.Context
import android.content.Intent
import org.autojs.autojs.AutoJs
import org.autojs.autojs.execution.ExecutionConfig
import org.autojs.autojs.model.script.PathChecker
import org.autojs.autojs.script.JavaScriptFileSource
import org.autojs.autojs.script.ScriptSource
import org.autojs.autojs.script.SequenceScriptSource
import org.autojs.autojs.script.StringScriptSource
import org.autojs.autojs.util.WorkingDirectoryUtils
import org.json.JSONException
import org.json.JSONObject
import java.io.File

/**
 * Created by Stardust on 2017/4/1.
 * Modified by SuperMonster003 as of Mar 25, 2023.
 * Transformed by SuperMonster003 on Mar 25, 2023.
 */
object ScriptIntents {

    const val EXTRA_KEY_PATH = "path"
    const val EXTRA_KEY_PRE_EXECUTE_SCRIPT = "script"
    const val EXTRA_KEY_LOOP_TIMES = "loop"
    const val EXTRA_KEY_LOOP_INTERVAL = "interval"
    const val EXTRA_KEY_DELAY = "delay"
    const val EXTRA_KEY_JSON = "json"

    @JvmStatic
    fun isTaskerJsonObjectValid(json: JSONObject) = json.has(EXTRA_KEY_PATH) || json.has(EXTRA_KEY_PRE_EXECUTE_SCRIPT)

    @JvmStatic
    fun handleIntent(context: Context?, intent: Intent) {
        var path = getPath(intent)
        var script = intent.getStringExtra(EXTRA_KEY_PRE_EXECUTE_SCRIPT)

        if (intent.hasExtra(EXTRA_KEY_JSON)) {
            try {
                JSONObject(intent.getStringExtra(EXTRA_KEY_JSON)!!).let {
                    if (it.has(EXTRA_KEY_PATH)) {
                        path = it[EXTRA_KEY_PATH] as String
                    }
                    if (it.has(EXTRA_KEY_PRE_EXECUTE_SCRIPT)) {
                        script = it[EXTRA_KEY_PRE_EXECUTE_SCRIPT] as String
                    }
                }
            } catch (e: JSONException) {
                // Ignored.
            }
        }
        ExecutionConfig().let { config ->
            config.apply {
                delay = intent.getLongExtra(EXTRA_KEY_DELAY, 0)
                loopTimes = intent.getIntExtra(EXTRA_KEY_LOOP_TIMES, 1)
                interval = intent.getLongExtra(EXTRA_KEY_LOOP_INTERVAL, 0)
                config.workingDirectory = WorkingDirectoryUtils.path
                setArgument("intent", intent)
            }
            var source: ScriptSource? = null
            when {
                path == null && script != null -> source = StringScriptSource(script)
                path != null && PathChecker(context).checkAndToastError(path) -> {
                    JavaScriptFileSource(path).let { fss: JavaScriptFileSource ->
                        source = script?.let { SequenceScriptSource(fss.name, StringScriptSource(it), fss) } ?: fss
                    }
                    File(path!!).parent?.let { config.workingDirectory = it }
                }
            }
            source?.let { AutoJs.instance.scriptEngineService.execute(it, config) }
        }
    }

    private fun getPath(intent: Intent) = intent.data?.path ?: intent.getStringExtra(EXTRA_KEY_PATH)

}